### PR TITLE
Proper invalidation of OGR sources on sync to disc

### LINKED ISSUE #39574: Cannot edit FlatGeobuf backed layer
While FlatGeobuf do not support random writes it certainly supports writing, and can be exported to etc but for some reason QGIS disables the edit tools for FlatGeobuf. I would presume it should work by keeping edits in memory and rewriting the full file on save, which I would also presume is how it works with fx. GeoJSON.

I've done some initial investigation on why I see this behavior, but have unable to find the cause as of yet.

### PR DESCRIPTION
**Pull request by bjornharrtell**
## Description

Aims to fix #39574 and #35144 using guidance from @jef-n and @rouault, thanks to you both.

See https://github.com/OSGeo/gdal/issues/3157 for background.

If this is a good fix, I think it should be considered a candidate for backport.

I have not yet verified if this actually resolves the issues as I currently lack a Windows development setup for GDAL + QGIS. This PR was done with a Ubuntu 20.04 based development setup and I've verified to the best of my ability that the changes work and have no adverse effects there.

Let me know if this needs to be verified on a Windows setup to be considered to merge, otherwise I will follow up with testing and verification on a nightly Windows build.

### CODE REVIEW

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -2544,11 +2544,10 @@ bool QgsOgrProvider::changeAttributeValues( const QgsChangedAttributesMap &attr_
   if ( mTransaction )
     mTransaction->dirtyLastSavePoint();
 
-  if ( mOgrLayer->SyncToDisk() != OGRERR_NONE )
+  if ( syncToDisc() == false )
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:2547]:
```suggestion
  if ( !syncToDisc() )
```

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 2]
```diff
@@ -2655,11 +2654,10 @@ bool QgsOgrProvider::changeGeometryValues( const QgsGeometryMap &geometry_map )
...
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:2657]:
```suggestion
  if ( !syncToDisc() )
```

**Review by rouault — COMMENTED** [review]
LGTM pending my stylistic remarks. In the absence of CI tests for Windows, I'd say let's merge this so that you can check with the next nightly Windows build if it is OK, and then backport it.

**Comment by bjornharrtell** [comment]
Thanks @rouault, accepted the stylistic changes, rebased and force pushed.

**Comment by bjornharrtell** [comment]
Happy to report that with the latest nightly QGIS build for Windows I cannot reproduce the issue any longer and edits works fine.

**Comment by bjornharrtell** [comment]
Backported with https://github.com/qgis/QGIS/pull/39949 but I see there is a bot that can do it somehow?!

**Pull request merged**